lib/federation/graphql-federation.factory.ts

```diff
@@ -132,9 +135,28 @@ export class GraphQLFederationFactory {
   ): GraphQLSchema {
     return transformSchema(executableSchema, (type) => {
       if (isUnionType(type) && type.name !== '_Entity') {
-        type.resolveType = (autoGeneratedSchema.getType(
+        const autogeneratedUnionType = autoGeneratedSchema.getType(
           type.name,
-        ) as GraphQLUnionType).resolveType;
+        ) as GraphQLUnionType;
+        type.resolveType = async (
+          value: any,
+          context: any,
+          info: GraphQLResolveInfo,
+          abstractType: GraphQLAbstractType,
+        ) => {
+          const result = await autogeneratedUnionType.resolveType(
+            value,
+            context,
+            info,
+            abstractType,
+          );
+          if (isString(result)) {
+            return result;
+          }
+          // for some reason this doesnt work if we return the resolved
+          // type object, but does if we return just the name
+          return result.name;
```

Comment:
@kamilmysliwiec this is the only difference between the default `resolveType` method created by `ResolveTypeFactory.getResolveTypeFunction`. I originally tried to do it the same way you did but the E2E search query test in `tests/e2e/code-first-federation.spec.ts` still failed with that method.